src/algebra/category/FinVect.lean

```diff
@@ -0,0 +1,165 @@
+/-
+Copyright (c) 2021 Jakob von Raumer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jakob von Raumer
+-/
+import category_theory.monoidal.rigid
+import linear_algebra.tensor_product_basis
+import linear_algebra.coevaluation
+import .Module.monoidal
+
+/-!
+# The category of finite dimensional vector spaces
+
+This introduces `FinVect K`, the category of finite dimensional vector spaces on a field `K`.
+It is implemented as a full subcategory on a subtype of  `Module K`.
+We first create the instance as a category, then as a monoidal category and then as a rigid monoidal
+category.
+
+## Future work
+
+* Show that `FinVect K` is a symmetric monoidal category.
+
+-/
+noncomputable theory
+
+open category_theory Module.monoidal_category
+open_locale classical big_operators
+
+universes u
+
+variables (K : Type u) [field K]
+
+/-- Define `FinVect` as the subtype of `Module.{u} K` of finite dimensional vector spaces. -/
+@[derive [category, has_coe_to_sort]]
+def FinVect := { V : Module.{u} K // finite_dimensional K V }
+
+namespace FinVect
+
+instance finite_dimensional (V : FinVect K): finite_dimensional K V := V.prop
+
+instance has_coe_to_fn (V W : FinVect K) : has_coe_to_fun (V ⟶ W) :=
+  ⟨λ _, V → W, λ f, f.to_fun⟩
+
+instance : inhabited (FinVect K) := ⟨⟨Module.of K K, finite_dimensional.finite_dimensional_self K⟩⟩
+
+instance monoidal_category : monoidal_category (FinVect K) :=
+@monoidal_category.full_monoidal_subcategory _ _ Module.monoidal_category
+ (λ (V : Module K), finite_dimensional K V)
+ (by { exact finite_dimensional.finite_dimensional_self K})
+ (λ X Y hX hY, @finite_dimensional_tensor_product K X Y _ _ _ _ _ hX hY)
+
+variables (V : FinVect K)
+
+/-- The dual module is the dual in the rigid monoidal category `FinVect K`. -/
+def FinVect_dual : FinVect K :=
+⟨Module.of K (module.dual K V), subspace.module.dual.finite_dimensional⟩
+
+open category_theory.monoidal_category
+
+/-- The coevaluation map is defined in `linear_algebra.coevaluation`. -/
+def FinVect_coevaluation : 𝟙_ (FinVect K) ⟶ V ⊗ (FinVect_dual K V) :=
+by { haveI := V.prop, change _ →ₗ[K] _, apply coevaluation K V }
+
+lemma FinVect_coevaluation_apply_one : FinVect_coevaluation K V (1 : K) =
+   ∑ (i : basis.of_vector_space_index K V),
+    (basis.of_vector_space K V) i ⊗ₜ[K] (basis.of_vector_space K V).coord i :=
+by { simp only [FinVect_coevaluation], apply coevaluation_apply_one K V }
```

Comment:
```suggestion
by { apply coevaluation_apply_one K V }
```
Curiously changing the `apply` to `exact` really slows this down, and I would still like to get to the bottom of that!!